### PR TITLE
Add initial __kconfig support with LINUX_KERNEL_VERSION.

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

This PR is a first step toward `__kconfig` support which is discussed in #698.
When a `.kconfig` data section is detected, a `.kconfig` map is created, its content will be initialized depending on the `CONFIG_*` values.
During relocation, variable inside the `.kconfig` data section will have their offset changed to reflect their value inside the `.kconfig` map.

For the moment, it only enables using `LINUX_KERNEL_VERSION`. The `.kconfig` map will be updated to contain the result of `internal.KernelVersion()`.

I tested it with some basic eBPF program I wrote:

```bash
root@81d71157c164:/mnt/cilium-ebpf-test-kconfig# more bpf/test.bpf.c 
// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_core_read.h>

extern int LINUX_KERNEL_VERSION __kconfig;

SEC("tp_btf/block_rq_insert")
int insert(u64 *ctx)
{
        bpf_printk("LINUX_KERNEL_VERSION: %d\n", LINUX_KERNEL_VERSION);

        return 0;
}

char LICENSE[] SEC("license") = "GPL";
root@81d71157c164:/mnt/ebpf# go run ./...
# Switch to other terminal
$ sudo cat /sys/kernel/debug/tracing/trace_pipe                                                               
 .../2-579     [006] d.... 11213.059870: bpf_trace_printk: LINUX_KERNEL_VERSION: 331584
```

Sadly, I was not able to make the added `TestLinuxKernelVersion()` works because the ELF misses BTF and I need `btfSpec` to not be `nil` to check data section.
I think I miss some steps in the ELF creation.

If you see any way to improve this PR, feel free to share.

Best regards and thank you in advance.

### CODE REVIEW

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:140]: This is more or less how I would do this:

```go
var ds *btf.Datasec
if err := ec.btf.TypeByName(".kconfig", &ds); err != nil && !errors.Is(err, btf.ErrNotFound) {
	return nil, err
}

data := make([]byte, 0, ds.Size)
for _, vsi := range ds.Vars {
	v := vsi.Type

	s, err := btf.Alignof(v)
	if err != nil {
		return nil, fmt.Errorf("getting size of kconfig variable %s: %w", v.TypeName(), err)
	}
	d := make([]byte, s)

	switch n := v.TypeName(); n {
	case "LINUX_KERNEL_VERSION":
		if s < 4 {
			return nil, fmt.Errorf("kconfig variable %s must be at least u32, got %d", n, s)
		}
		kv, err := internal.KernelVersion()
		if err != nil {
			return nil, fmt.Errorf("getting kernel version: %w", err)
		}
		internal.NativeEndian.PutUint32(d, kv.Kernel())

	default:
		return nil, fmt.Errorf("unsupported kconfig variable: %s", n)
	}

	data = append(data, d...)
}
```

**Review by ti-mo — CHANGES_REQUESTED** [review]
Gave this a first round. Running the tests doesn't work for me.

File: btf/btf.go
Comment (ti-mo) [reply btf/btf.go:546]: No need, this method is not necessary.

File: testdata/linux_kernel_version.c
Comment (ti-mo) [reply testdata/linux_kernel_version.c:1]: `strip` doesn't throw away the .BTF section and friends, and clearly `loader.c` ELFs carry BTF info:
```
λ  ebpf francis/ebpf-kconfig ✓ bpftool btf dump file testdata/loader-clang-14-el.elf
[1] PTR '(anon)' type_id=3
[2] INT 'int' size=4 bits_offset=0 nr_bits=32 encoding=SIGNED
...
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:30]:
```suggestion
	maps     map[string]*MapSpec
```
Go maps are already pointer types and nil by default, no need to take out pointers to them.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:144]: Please factor `maps` out from being a parameter on the methods above (in a separate commit) and initialize it when allocating the `elfCode` above. This is a no go, sorry.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1133]: These kinds of comments are useful for holding your train of thought, but are a net negative on the clarity of the code, please avoid them.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1139]: The code in `loadDataSections` already guarantees this is a ds, please drop the check to lower the verbosity.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:590]: You already perform this check in `populateDataSections()`. `v := vsi.Type.(*btf.Var)` is fine here.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:589]: For consistency with the population counterpart:
```suggestion
				for _, vsi := range m.Value.(*btf.Datasec).Vars {
```
Also, m.Value is guaranteed to be a ds by preceding code.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:587]:
```suggestion
		if m := ec.maps[".kconfig"]; m != nil {
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:585]: This comment is not so useful.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1123]: Not a useful check.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1127]:
```suggestion
	if err := ec.populateKconfigSection(); err != nil {
		return fmt.Errorf(".kconfig: %w", err)
	}
	
	return nil
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1147]: This will make debugging easier on a production build in the wild:
```suggestion
			return fmt.Errorf("variable %s: unexpected type %T", vsi.Type.TypeName(), vsi.Type)
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1152]: For consistency with the above:
```suggestion
			return fmt.Errorf("variable %s: getting size: %w", v.Name, err)
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1159]: Error wrapping will already annotate this with `.kconfig`.
```suggestion
				return fmt.Errorf("variable %s must be at least u32, got %d", n, s)
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1168]:
```suggestion
			return fmt.Errorf("unsupported variable: %s", n)
```

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1180]: Which problems?

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1178]: This variable is only used once.

File: elf_reader.go
Comment (ti-mo) [inline elf_reader.go:1184]: Let's not initialize ValueSize in `loadDataSections` if we need to fix it up here anyway.

**Review by ti-mo — CHANGES_REQUESTED** [review]

**Comment by brycekahle** [comment]
@lmb @ti-mo We are interested in getting this over the finish line too, so if I can help, please let me know.

**Comment by ti-mo** [comment]
> I did not abandon this work, I was just waiting for some feedback.

Sorry Francis, that's some bad communication on my part. You've addressed all my feedback nicely and things are in a good shape! I hadn't taken the time to look at this a bit closer up, so I discovered a few unknown unknowns this week. I was hoping I could collaborate by adding a few patches underneath or on top before merging this, that's all. Sorry for the misunderstanding, I wasn't implying you abandoned the work at all. :blush: 

> What is the purpose of asm_relocation()? Is this only proper to cilium/ebpf because when implementing this I got a lot of inspiration from what is done in libbpf.

Correct, this is an old hack that was relied on by some users to implement static data before the kernel had support for pseudo/frozen maps like .(ro)data. There are still kernels out there that don't support them, so we're not going to drop support for this right now.

We have a similar issue for adding kfunc support, since we've always permitted forward declarations in C (see `fwd_decl.c`). This allows the caller to append to `Instructions` with `.WithSymbol("fwd")` after loading the ELF, but before loading the prog into the kernel, to inject behaviour at runtime without recompiling.

Had a discussion with Lorenz about this, we worked out some of the compatibility concerns I had around `asm_relocation`:
- If .kconfig is present, all relos against NOTYPE/GLOBAL/UND symbols must be executed against .kconfig. Any such symbols that don't appear in .kconfig are an error.
- We won't support __weak in this PR, that can be done later (since only LINUX_KERNEL_VERSION atm)

We're still debating whether or not we should expose .kconfig's MapSpec.Contents in the CollectionSpec, or if we'll populate it during map load.

**Comment by ti-mo** [comment]
> How do you plan to add the patches?

By default, maintainers of a repo are allowed to push to PR branches in forks, unless explicitly disabled. The sidebar of this PR says `Maintainers are allowed to edit this pull request.` for me, but I don't seem to manage to push to it. Could you check if this isn't disabled on your end? No need to create other PRs, let's keep the discussion centered here.

> So, you mean to put in .kconfig even symbols which are not marked as __kconfig?

No, I mean we wouldn't support both __kconfig and inline asm relocations (like `MY_CONST` in `asm_relocation()`) in the same ELF. If a .kconfig map is present, all NOTYPE/GLOBAL/UND symbols must exist in it. If no .kconfig map is present and we need to relocate a MY_CONST, we just call `ins.WithReference()` and call it a day. The linker will complain later if the caller didn't fix up the instruction and didn't remove the reference.

**Comment by ti-mo** [comment]
@eiffel-fl Sorry, on PTO this week. Just gave this another try (this is after `gh pr checkout eiffel-fl:francis/ebpf-kconfig` and reworking existing commits and prepending one commit to the existing two:

```
λ  ebpf francis/ebpf-kconfig ✓ git push -f
ERROR: Permission to eiffel-fl/ebpf.git denied to ti-mo.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Do you have any kind of branch protection set up on your fork? There's a branch protection setting that controls force-pushes, but not sure if it's on or off by default.

**Comment by ti-mo** [comment]
@eiffel-fl Could you try re-forking from `cilium/ebpf` instead, and recreating this PR using my branch at https://github.com/ti-mo/ebpf/tree/tb/kconfig? I've added some code that fixes the alignment of the .kconfig Datasec and moved populating .kconfig to happen at map load time instead of ELF load time. Otherwise, missing kconfig values would prevent the ELF from loading, leaving the user with no way to fix things up.

Please close this PR when you've recreated it.

**Pull request closed without merging**
